### ISSUE TITLE
IOS Dialog popup is requesting every time while start app, if was choose not the waited "#locationAuthorizationRequest" 

### ISSUE THREAD

**Issue by javurekv** [transistorsoft/cordova-background-geolocation-lt#522]
## Your Environment
* Plugin version: 2.9.0 (2.8.5 as well)
* Platform: iOS
* OS version: 11
* Device manufacturer / model: all ios device from 5s
* Cordova version (`cordova -v`): 7.1.0
* Cordova platform version (`cordova platform ls`): ios 4.5.1
* Plugin config

            desiredAccuracy: 0,
            distanceFilter: 200,
            stationaryRadius: 25,
            activityRecognitionInterval: 10000,
            debug: false,
            stopTimeout: 5,
            stopOnTerminate: false,
            startOnBoot: true,
            locationAuthorizationRequest: 'Always',
            heartbeatInterval: 60,
            logLevel: this.bgGeo.LOG_LEVEL_VERBOSE,
            method: "POST",
            autoSync: true,
            url: `${environment.api_endpoint}/users/track/background`,
            httpRootProperty:".",
            locationTemplate: '{ "longitude":<%= longitude %>, "latitude":<%= latitude %> }',
            headers: {
                'Content-Type': 'application/json',
                'Accept': 'application/prs.app.v1+json',
                'Authorization': `Bearer ${token}`
            },
        }, (state) => {
            this.bgGeo.start();
        });

and 

        this.bgGeo.on('providerchange', (provider) => {
            switch(provider.status) {
                case this.bgGeo.AUTHORIZATION_STATUS_DENIED:
                this.bgGeo.setConfig({locationAuthorizationRequest: 'Always'});
                break;
                case this.bgGeo.AUTHORIZATION_STATUS_ALWAYS:
                this.bgGeo.setConfig({locationAuthorizationRequest: 'Always'});
                break;
                case this.bgGeo.AUTHORIZATION_STATUS_WHEN_IN_USE:
                this.bgGeo.setConfig({locationAuthorizationRequest: 'WhenInUse'});
                break;
            }
        });

## Expected Behavior
If user choose "when use" or "always" - I dont want see the geolocation permission popup anymore. If user choose "cancel" - should show the popup again. 
In other words, "when use" and "always" are good bouth for us.

## Actual Behavior
If user has set location permission not the same like #locationAuthorizationRequest - we see permission popup.

Do you know some way to get it fix?

Thank you.

**Comment by christocracy** [comment]
```javascript
case this.bgGeo.AUTHORIZATION_STATUS_DENIED:
  this.bgGeo.stop();  // <-- Turn it off to prevent popups
  break;
case this.bgGeo.AUTHORIZATION_STATUS_ALWAYS:
  this.bgGeo.setConfig({locationAuthorizationRequest: 'Always'}, (state) => {
    if (!state.enabled) { this.bgGeo.start(); }  // <-- turn it back on.
  });
case this.bgGeo.AUTHORIZATION_STATUS_WHEN_IN_USE:
  this.bgGeo.setConfig({locationAuthorizationRequest: 'WhenInUse'}, (state) => {
    if (!state.enabled) { this.bgGeo.start(); }  // <-- turn it back on.
  });
  break;
```

**Comment by christocracy** [comment]
Update?

**Comment by javurekv** [comment]
@christocracy No it does not work. Popup shows again while start when I choose "when use" and does not show when I choose "Always" or "Never". But I want see the popup again only when was choose "Never".
My code:

        this.bgGeo.on('providerchange', (provider) => {
            switch(provider.status) {
                case this.bgGeo.AUTHORIZATION_STATUS_DENIED:
                this.bgGeo.stop();
                break;
                case this.bgGeo.AUTHORIZATION_STATUS_ALWAYS:
                this.bgGeo.setConfig({locationAuthorizationRequest: 'Always'}, (state) => {
                    if (!state.enabled) {
                        this.bgGeo.start();
                    }
                });
                break;
                case this.bgGeo.AUTHORIZATION_STATUS_WHEN_IN_USE:
                this.bgGeo.setConfig({locationAuthorizationRequest: 'WhenInUse'}, (state) => {
                    if (!state.enabled) {
                        this.bgGeo.start();
                    }
                });
                break;
            }
        });
        this.bgGeo.configure({
            desiredAccuracy: 0,
            distanceFilter: 0,
            locationUpdateInterval : 15 * 60 * 1000,
            deferTime : 15 * 60 * 1000,
            activityRecognitionInterval : 15 * 60 * 1000,
            stationaryRadius: 25,
            debug: false,  // <-- Debug sounds & notifications.
            stopTimeout: 5,
            stopOnTerminate: false,
            startOnBoot: true,
            locationAuthorizationRequest: 'Always',
            heartbeatInterval: 60,
            logLevel: this.bgGeo.LOG_LEVEL_VERBOSE,
            method: "POST",
            autoSync: true,
            url: `${environment.api_endpoint}/users/track/background`,
            httpRootProperty:".",
            locationTemplate: '{ "longitude":<%= longitude %>, "latitude":<%= latitude %> }',
            headers: {
                'Content-Type': 'application/json',
                'Accept': 'application/prs.app.v1+json',
                'Authorization': `Bearer ${token}`
            },
        }, (state) => {
            if (!state.enabled) {
                this.bgGeo.start(); 
            }
        });

**Comment by christocracy** [comment]
In that case, it's even simpler:  just don't `#stop` the plugin when `DENIED` is detected.

As long as the plugin is `#start`ed, it will complain when the user selects anything other than `locationAuthorizationRequest`.

```javascript
switch(provider.status) {
      case this.bgGeo.AUTHORIZATION_STATUS_ALWAYS:
        // Change expectation to 'Always'
        this.bgGeo.setConfig({locationAuthorizationRequest: 'Always'});
        break;
      case this.bgGeo.AUTHORIZATION_STATUS_WHEN_IN_USE:
        // Change expectation to 'WhenInUse`
        this.bgGeo.setConfig({locationAuthorizationRequest: 'WhenInUse'});
        break;
    }
```

**Comment by javurekv** [comment]
The code did not help. One strange thing in logs while start. I see twice the blocks
`╔═══════════════════════════════════════════════════════════`
`║ -[Settings load:] `

One time with `locationAuthorizationRequest = Always;`
Next time with `locationAuthorizationRequest = WhenInUse;`

**Comment by christocracy** [comment]
> I see twice the blocks

- Once for the initial `#configure`.  The `#configure` method then fires the `providerchange` event
- Second time is your `providerchange` handler executing `#setConfig`

**Comment by javurekv** [comment]
It happened after: install app -> choose "When use" -> close app -> open app.

So I should get only one the block with `locationAuthorizationRequest = WhenInUse.`

**Comment by javurekv** [comment]
After start app always get the popup
![2017-11-13 16 02 45](https://user-images.githubusercontent.com/7250080/32731097-940011ee-c891-11e7-9f49-2dea32ef81d3.png)

**Comment by christocracy** [comment]
In your initial `#configure`, you're providing a hard-coded `Always`:
```javascript

this.bgGeo.configure({
  locationAuthorizationRequest: 'Always',  // <--  You need to put this in localStorage or something
}, (state) => {
  // providerchange has fired at this point.
});

**Comment by christocracy** [comment]
You could also do this:
```javascript
this.bgGeo.getState((config) => {
  // this is the last settings the plugin was provided.
  console.log('- last plugin state': config);
  // provide overrides here
  config.logLevel = this.bgGeo.LOG_LEVEL_VERBOSE;
  config.distanceFilter = 0;
  .
  .
  this.bgGeo.configure(config, (state) => {

  });
});
```

**Comment by javurekv** [comment]
How I know `locationAuthorizationRequest:` by default is 'Always', so it's the only correct way.
But I get the popup after first open app and choose "When use" or "Never". After this all works well.

**Comment by christocracy** [comment]
Ok.  Throw `#stop` in there.  This will ensure the 2nd popup is hidden.  You *will* see the 2nd popup briefly appear then disappear but I have a solution for that coming.

```      
    switch(provider.status) {
      case bgGeo.AUTHORIZATION_STATUS_ALWAYS:
        // Change expectation to 'Always'
        bgGeo.stop();  // <-- stop hides all popups.
        bgGeo.setConfig({locationAuthorizationRequest: 'Always'}, (state) => {
          bgGeo.start();
        });
        break;
      case bgGeo.AUTHORIZATION_STATUS_WHEN_IN_USE:
        // Change expectation to 'WhenInUse`
        bgGeo.stop();  // <-- stop hides all popups.
        bgGeo.setConfig({locationAuthorizationRequest: 'WhenInUse'}, (state) => {
          bgGeo.start();
        });
        break;
    }
```

**Comment by javurekv** [comment]
It works well, thank you! 
We are waiting for the solution.
